go/libraries/doltcore/sqle/dolt_diff_summary_table_function.go

```diff
@@ -0,0 +1,503 @@
+	if len(expression) < 2 || len(expression) > 3 {
+		return nil, sql.ErrInvalidArgumentNumber.New(ds.FunctionName(), "2 or 3", len(expression))
+	}
+
+	for _, expr := range expression {
+		if !expr.Resolved() {
+			return nil, ErrInvalidNonLiteralArgument.New(ds.FunctionName(), expr.String())
+		}
+	}
+
+	ds.fromCommitExpr = expression[0]
+	ds.toCommitExpr = expression[1]
+	if len(expression) == 3 {
+		ds.tableNameExpr = expression[2]
+	}
+
+	// validate the expressions
+	if !sql.IsText(ds.fromCommitExpr.Type()) {
+		return nil, sql.ErrInvalidArgumentDetails.New(ds.FunctionName(), ds.fromCommitExpr.String())
+	}
+
+	if !sql.IsText(ds.toCommitExpr.Type()) {
+		return nil, sql.ErrInvalidArgumentDetails.New(ds.FunctionName(), ds.toCommitExpr.String())
+	}
+
+	if ds.tableNameExpr != nil {
+		if !sql.IsText(ds.tableNameExpr.Type()) {
+			return nil, sql.ErrInvalidArgumentDetails.New(ds.FunctionName(), ds.tableNameExpr.String())
+		}
+	}
+
+	return ds, nil
+}
+
+// RowIter implements the sql.Node interface
+func (ds *DiffSummaryTableFunction) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	fromCommitVal, toCommitVal, tableName, err := ds.evaluateArguments()
+	if err != nil {
+		return nil, err
+	}
+
+	sqledb, ok := ds.database.(Database)
+	if !ok {
+		panic(fmt.Sprintf("unexpected database type: %T", ds.database))
+	}
+
+	sess := dsess.DSessFromSess(ctx.Session)
+	fromRoot, _, err := sess.ResolveRootForRef(ctx, sqledb.Name(), fromCommitVal)
+	if err != nil {
+		return nil, err
+	}
+
+	toRoot, _, err := sess.ResolveRootForRef(ctx, sqledb.Name(), toCommitVal)
+	if err != nil {
+		return nil, err
+	}
+
+	deltas, err := diff.GetTableDeltas(ctx, fromRoot, toRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// If tableNameExpr defined, return a single table diff summary result
+	if ds.tableNameExpr != nil {
+		delta := findMatchingDelta(deltas, tableName)
+		diffSum, hasDiff, err := getDiffSummaryNodeFromDelta(ctx, delta, fromRoot, toRoot, tableName)
+		if err != nil {
+			return nil, err
+		}
+		if !hasDiff {
+			return NewDiffSummaryTableFunctionRowIter([]diffSummaryNode{}), nil
+		}
+		return NewDiffSummaryTableFunctionRowIter([]diffSummaryNode{diffSum}), nil
+	}
+
+	var diffSummaries []diffSummaryNode
+	for _, delta := range deltas {
+		tblName := delta.ToName
+		if tblName == "" {
+			tblName = delta.FromName
+		}
+		diffSum, hasDiff, err := getDiffSummaryNodeFromDelta(ctx, delta, fromRoot, toRoot, tblName)
+		if err != nil {
+			return nil, err
+		}
+		if hasDiff {
+			diffSummaries = append(diffSummaries, diffSum)
+		}
+	}
+
+	return NewDiffSummaryTableFunctionRowIter(diffSummaries), nil
+}
+
+// evaluateArguments returns fromCommitValStr, toCommitValStr and tableName.
+// It evaluates the argument expressions to turn them into values this DiffTableFunction
+// can use. Note that this method only evals the expressions, and doesn't validate the values.
```

Comment:
Wonderful function docs! Those string return params would be much harder for people to figure out without these nice docs. 🙌